File: cb/api/crumb.py
from __future__ import annotations
"""CSRF crumb fetcher for CloudBees/Jenkins POST requests."""

import time
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from cb.api.client import CloudBeesClient

_crumb_cache: dict = {}  # keyed by client base_url
_CRUMB_TTL = 300  # 5 minutes


def get_crumb(client: "CloudBeesClient") -> Optional[dict]:
    """
    Fetch (or return cached) CSRF crumb for this CloudBees server.
    Returns {"field": "Jenkins-Crumb", "value": "abc123"} or None.
    """
    key = client.base_url
    entry = _crumb_cache.get(key)
    if entry and time.time() < entry["expires_at"]:
        return entry["crumb"]

    try:
        data = client._request("GET", "/crumbIssuer/api/json")
        if data and "crumb" in data:
            crumb = {
                "field": data.get("crumbRequestField", "Jenkins-Crumb"),
                "value": data["crumb"],
            }
            _crumb_cache[key] = {
                "crumb": crumb,
                "expires_at": time.time() + _CRUMB_TTL,
            }
            return crumb
    except Exception:
        pass  # Server may not have CSRF protection enabled
    return None


def invalidate_crumb(base_url: str) -> None:
    """Remove cached crumb for a server (call after 403 response)."""
    _crumb_cache.pop(base_url, None)

File: src/core/api/crumb.ts
/**
 * CSRF crumb fetcher — 1:1 port of legacy/cb/api/crumb.py
 *
 * Module-level Map cache keyed by baseUrl, 5-minute TTL.
 */

/** Shape returned by getCrumb and stored in cache. */
export interface CrumbData {
  field: string;
  value: string;
}

interface CrumbCacheEntry {
  crumb: CrumbData;
  expiresAt: number; // ms since epoch
}

/** Internal interface the crumb module needs from the client. */
export interface CrumbClient {
  readonly baseUrl: string;
  /** Low-level unauthenticated-crumb-free GET used only by crumb.ts. */
  _rawGet(path: string): Promise<unknown>;
}

const _crumbCache = new Map<string, CrumbCacheEntry>();
const CRUMB_TTL_MS = 300 * 1000; // 5 minutes in milliseconds

/**
 * Fetch (or return cached) CSRF crumb for this CloudBees server.
 * Returns { field, value } or null if server has CSRF disabled.
 */
export async function getCrumb(client: CrumbClient): Promise<CrumbData | null> {
  const key = client.baseUrl;
  const entry = _crumbCache.get(key);
  if (entry && Date.now() < entry.expiresAt) {
    return entry.crumb;
  }

  try {
    const data = await client._rawGet("/crumbIssuer/api/json");
    if (
      data !== null &&
      typeof data === "object" &&
      "crumb" in data &&
      typeof (data as Record<string, unknown>)["crumb"] === "string"
    ) {
      const rec = data as Record<string, unknown>;
      const crumb: CrumbData = {
        field:
          typeof rec["crumbRequestField"] === "string"
            ? rec["crumbRequestField"]
            : "Jenkins-Crumb",
        value: rec["crumb"] as string,
      };
      _crumbCache.set(key, { crumb, expiresAt: Date.now() + CRUMB_TTL_MS });
      return crumb;
    }
  } catch {
    // Server may not have CSRF protection enabled — return null
  }
  return null;
}

/** Remove cached crumb for a server (call after 403 response). */
export function invalidateCrumb(baseUrl: string): void {
  _crumbCache.delete(baseUrl);
}
